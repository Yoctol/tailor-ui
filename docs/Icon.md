---
id: icon
title: Icon
---

Semantic vector graphics.

## Examples

```js
import { Icon } from 'tailor-ui';
import { TiFlowChildren } from 'react-icons/ti';
import { MdEdit, MdComment, MdInput, MdInfo } from 'react-icons/md';
```

### Built-in

```jsx live
<>
  <Icon mx="2" type="line" />
  <Icon mx="2" type="messenger" />
  <Icon mx="2" type="robot" />
  <Icon mx="2" type="understood" />
  <Icon mx="2" type="kurator" />
  <Icon mx="2" type="touch" />
  <Icon mx="2" type="analytics" />
  <Icon mx="2" type="tags" />
  <Icon mx="2" type="settings" />
  <Icon mx="2" type="tutorial" />
  <Icon mx="2" type="info" />
  <Icon mx="2" type="success" />
  <Icon mx="2" type="warning" />
  <Icon mx="2" type="danger" />
</>
```

### With text

```jsx live
<Flex alignItems="center">
  This is a info icon <Icon ml="2" size="20" cursor="help" type={MdInfo} />
</Flex>
```

### Custom props

```jsx live
<>
  <Icon mx="2" fill="#00c300" type="line" />
  <Icon mx="2" fill="#0084ff" type="messenger" />
  <Icon mx="2" fill="primaryLight" size="32" type="understood" />
  <Icon mx="2" fill="primaryLight" size="32" type="kurator" />
  <Icon mx="2" fill="primaryLight" size="32" type="touch" />
  <Icon mx="2" fill="primaryLight" size="32" type="analytics" />
  <Icon mx="2" fill="primaryLight" size="32" type="tags" />
  <Icon mx="2" fill="primaryLight" size="32" type="settings" />
  <Icon mx="2" fill="primaryDark" size="48" type={MdEdit} />
  <Icon mx="2" fill="primaryDark" size="48" type={MdComment} />
  <Icon mx="2" fill="primaryDark" size="48" type={MdInput} />
  <Icon mx="2" fill="primaryDark" size="48" type={TiFlowChildren} />
</>
```

### react-icons

```jsx live
<>
  <Icon mx="2" type={MdEdit} />
  <Icon mx="2" type={MdComment} />
  <Icon mx="2" type={MdInput} />
  <Icon mx="2" type={TiFlowChildren} />
</>
```

### cursor

```jsx live
<>
  <Icon mx="2" type={MdEdit} />
  <Icon mx="2" cursor="pointer" type={MdComment} />
  <Icon mx="2" cursor="grab" type={MdInput} />
  <Icon mx="2" cursor="help" type={TiFlowChildren} />
</>
```

## API
