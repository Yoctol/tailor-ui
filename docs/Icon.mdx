---
name: Icon
menu: Components
route: /components/icon
---

import { PropsTable, Playground } from 'docz';
import { TiFlowChildren } from 'react-icons/ti';
import { MdEdit, MdComment, MdInput, MdInfo } from 'react-icons/md';
import { Flex, Icon } from 'tailor-ui';

# Icon

Semantic vector graphics.

## Examples

```js
import { Icon } from 'tailor-ui';
import { TiFlowChildren } from 'react-icons/ti';
import { MdEdit, MdComment, MdInput, MdInfo } from 'react-icons/md';
```

### Built-in

<Playground>
  <Icon mx="2" type="line" />
  <Icon mx="2" type="messenger" />
  <Icon mx="2" type="robot" />
  <Icon mx="2" type="understood" />
  <Icon mx="2" type="kurator" />
  <Icon mx="2" type="touch" />
  <Icon mx="2" type="analytics" />
  <Icon mx="2" type="tags" />
  <Icon mx="2" type="settings" />
  <Icon mx="2" type="tutorial" />
  <Icon mx="2" type="info" />
  <Icon mx="2" type="success" />
  <Icon mx="2" type="warning" />
  <Icon mx="2" type="danger" />
</Playground>

### With text

<Playground>
  <Flex alignItems="center">
    This is a info icon <Icon ml="2" size="20" cursor="help" type={MdInfo} />
  </Flex>
</Playground>

### Custom props

<Playground>
  <Icon mx="2" fill="#00c300" type="line" />
  <Icon mx="2" fill="#0084ff" type="messenger" />
  <Icon mx="2" fill="primaryLight" size="32" type="understood" />
  <Icon mx="2" fill="primaryLight" size="32" type="kurator" />
  <Icon mx="2" fill="primaryLight" size="32" type="touch" />
  <Icon mx="2" fill="primaryLight" size="32" type="analytics" />
  <Icon mx="2" fill="primaryLight" size="32" type="tags" />
  <Icon mx="2" fill="primaryLight" size="32" type="settings" />
  <Icon mx="2" fill="primaryDark" size="48" type={MdEdit} />
  <Icon mx="2" fill="primaryDark" size="48" type={MdComment} />
  <Icon mx="2" fill="primaryDark" size="48" type={MdInput} />
  <Icon mx="2" fill="primaryDark" size="48" type={TiFlowChildren} />
</Playground>

### react-icons

<Playground>
  <Icon mx="2" type={MdEdit} />
  <Icon mx="2" type={MdComment} />
  <Icon mx="2" type={MdInput} />
  <Icon mx="2" type={TiFlowChildren} />
</Playground>

### cursor

<Playground>
  <Icon mx="2" type={MdEdit} />
  <Icon mx="2" cursor="pointer" type={MdComment} />
  <Icon mx="2" cursor="grab" type={MdInput} />
  <Icon mx="2" cursor="help" type={TiFlowChildren} />
</Playground>

## API

<PropsTable of={Icon} />
