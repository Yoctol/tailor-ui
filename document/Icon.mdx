---
name: Icon
menu: Components
route: /components/icon
---

import { PropsTable, Playground } from 'docz';
import { TiFlowChildren } from 'react-icons/ti';
import { MdEdit, MdComment, MdInput, MdInfo } from 'react-icons/md';
import { Flex, Icon } from 'tailor-ui';

// import \* as icons from 'icons';

# Icon

Semantic vector graphics.

<!-- ## Built-In Icon

<ul>
  {Object.keys(icons).map(icon => (
    <li key={icon}>{icon}</li>
  ))}
</ul> -->

## Examples

```js
import { Icon } from 'tailor-ui';
import { TiFlowChildren } from 'react-icons/ti';
import { MdEdit, MdComment, MdInput, MdInfo } from 'react-icons/md';
```

### Built-in

<Playground>
  <Icon mx={2} type="line" />
  <Icon mx={2} type="messenger" />
  <Icon mx={2} type="understood" />
  <Icon mx={2} type="kurator" />
  <Icon mx={2} type="touch" />
  <Icon mx={2} type="analytics" />
  <Icon mx={2} type="tags" />
  <Icon mx={2} type="settings" />
</Playground>

### With text

<Playground>
  <Flex alignItems="center">
    This is a info icon <Icon ml="2" size="20" cursor="help" type={MdInfo} />
  </Flex>
</Playground>

### Custom props

<Playground>
  <Icon mx={2} fill="#00c300" type="line" />
  <Icon mx={2} fill="#0084ff" type="messenger" />
  <Icon mx={2} fill="primaryLight" size={32} type="understood" />
  <Icon mx={2} fill="primaryLight" size={32} type="kurator" />
  <Icon mx={2} fill="primaryLight" size={32} type="touch" />
  <Icon mx={2} fill="primaryLight" size={32} type="analytics" />
  <Icon mx={2} fill="primaryLight" size={32} type="tags" />
  <Icon mx={2} fill="primaryLight" size={32} type="settings" />
  <Icon mx={2} fill="secondary" size={48} type={MdEdit} />
  <Icon mx={2} fill="secondary" size={48} type={MdComment} />
  <Icon mx={2} fill="secondary" size={48} type={MdInput} />
  <Icon mx={2} fill="secondary" size={48} type={TiFlowChildren} />
</Playground>

### react-icons

<Playground>
  <Icon mx={2} type={MdEdit} />
  <Icon mx={2} type={MdComment} />
  <Icon mx={2} type={MdInput} />
  <Icon mx={2} type={TiFlowChildren} />
</Playground>

### cursor

<Playground>
  <Icon mx={2} type={MdEdit} />
  <Icon mx={2} cursor="pointer" type={MdComment} />
  <Icon mx={2} cursor="grab" type={MdInput} />
  <Icon mx={2} cursor="help" type={TiFlowChildren} />
</Playground>

### loading

<Playground>
  <Icon size="40" type="loading" />
</Playground>

## API

<PropsTable of={Icon} />
