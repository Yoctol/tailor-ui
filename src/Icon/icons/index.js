export { default as line } from './line.svg';
export { default as messenger } from './messenger.svg';
export { default as understood } from './understood.svg';
export { default as kurator } from './kurator.svg';
export { default as touch } from './touch.svg';
export { default as analytics } from './analytics.svg';
export { default as settings } from './settings.svg';
