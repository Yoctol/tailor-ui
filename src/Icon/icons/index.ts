export { default as line } from './line';
export { default as messenger } from './messenger';
export { default as understood } from './understood';
export { default as kurator } from './kurator';
export { default as touch } from './touch';
export { default as analytics } from './analytics';
export { default as tags } from './tags';
export { default as settings } from './settings';
export { default as loading } from './loading';
