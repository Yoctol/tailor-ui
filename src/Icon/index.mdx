---
name: Icon
menu: General
route: /general/icon
---
import { PropsTable, Playground } from 'docz';
import { TiFlowChildren } from 'react-icons/lib/ti';
import { MdEdit, MdComment, MdInput } from 'react-icons/lib/md';

import Icon from './';

# Icon

## Properites

<PropsTable of={Icon} />

## Usage

```js
import { Icon } from '@yoctol/ui';
import { TiFlowChildren } from 'react-icons/lib/ti';
import { MdEdit, MdComment, MdInput } from 'react-icons/lib/md';
```

### Built-in

<Playground>
  <Icon mx={2} type="line" />
  <Icon mx={2} type="messenger" />
  <Icon mx={2} type="understood" />
  <Icon mx={2} type="kurator" />
  <Icon mx={2} type="touch" />
  <Icon mx={2} type="analytics" />
  <Icon mx={2} type="tags" />
  <Icon mx={2} type="settings" />
</Playground>


### Custom props

<Playground>
  <Icon mx={2} fill="#00c300" type="line" />
  <Icon mx={2} fill="#0084ff" type="messenger" />
  <Icon mx={2} fill="primaryLight" size={32} type="understood" />
  <Icon mx={2} fill="primaryLight" size={32} type="kurator" />
  <Icon mx={2} fill="primaryLight" size={32} type="touch" />
  <Icon mx={2} fill="primaryLight" size={32} type="analytics" />
  <Icon mx={2} fill="primaryLight" size={32} type="tags" />
  <Icon mx={2} fill="primaryLight" size={32} type="settings" />
  <Icon mx={2} fill="secondary" size={48} type={MdEdit} />
  <Icon mx={2} fill="secondary" size={48} type={MdComment} />
  <Icon mx={2} fill="secondary" size={48} type={MdInput} />
  <Icon mx={2} fill="secondary" size={48} type={TiFlowChildren} />
</Playground>

### react-icons

<Playground>
  <Icon mx={2} type={MdEdit} />
  <Icon mx={2} type={MdComment} />
  <Icon mx={2} type={MdInput} />
  <Icon mx={2} type={TiFlowChildren} />
</Playground>

### cursor

<Playground>
  <Icon mx={2} type={MdEdit} />
  <Icon
    mx={2}
    cursor="pointer"
    type={MdComment}
  />
  <Icon
    mx={2}
    cursor="grab"
    type={MdInput}
  />
  <Icon
    mx={2}
    cursor="help"
    type={TiFlowChildren}
  />
</Playground>

### loading

<Playground>
  <Icon size="40" type="loading" />
</Playground>
