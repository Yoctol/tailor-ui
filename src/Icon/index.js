import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { style, space } from 'styled-system';
import { prop } from 'styled-tools';

import * as icons from './icons';

const fill = style({
  prop: 'fill',
  cssProperty: 'fill',
  key: 'colors',
});

const size = css`
  width: ${prop('size')}px;
  height: ${prop('size')}px;

  svg {
    width: ${prop('size')}px;
    height: ${prop('size')}px;
  }
`;

const IconWrapper = styled.i`
  display: inline-block;
  cursor: ${prop('cursor')};

  svg {
    vertical-align: middle;

    ${fill};
  }

  ${size};
  ${space};
`;

IconWrapper.propTypes = {
  cursor: PropTypes.string.isRequired,
  ...size.propTypes,
  ...space.propTypes,
};

IconWrapper.defaultProps = {
  size: 24,
  fill: 'bodyFont',
};

const Icon = ({ type, ...otherProps }) => {
  let IconComponent = type;

  if (typeof type === 'string') {
    const BuiltInIcon = icons[type];
    if (BuiltInIcon) {
      IconComponent = BuiltInIcon;
    } else {
      return console.error('Built-in icon does not exists!');
    }
  }

  return (
    <IconWrapper {...otherProps} className="icon">
      <IconComponent />
    </IconWrapper>
  );
};

Icon.propTypes = {
  cursor: PropTypes.string,
  type: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
};

Icon.defaultProps = {
  cursor: 'default',
};

export default Icon;
