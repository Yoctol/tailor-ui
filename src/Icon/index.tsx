import React, { SFC } from 'react';
import { IconType as ReactIconsIconType } from 'react-icons';
import { SpaceProps, style, space as styledSpace } from 'styled-system';

import styled, { css } from 'utils/styled-components';

import * as icons from './icons';

interface Icons {
  [key: string]: any;
}

const styledFill = style({
  prop: 'fill',
  cssProperty: 'fill',
  key: 'colors',
});

type Size = string | number;

const styledSize = css<{ size: Size }>`
  width: ${p => p.size}px;
  height: ${p => p.size}px;
`;

export type IconWrapperProps = SpaceProps & {
  cursor?: string;
  size?: Size;
  fill?: string;
};

export const IconWrapper = styled<IconWrapperProps, 'i'>('i')`
  display: inline-flex;
  line-height: 1;
  cursor: ${p => p.cursor};

  svg {
    fill: ${p => p.theme.colors.primary};
    vertical-align: middle;
    ${p => p.theme.transition /* sc-declaration */};
  }

  &&& svg {
    ${styledFill};
    ${styledSize};
  }

  ${styledSpace};
`;

export type IconType = string | ReactIconsIconType;

export type IconProps = SpaceProps & {
  cursor?: string;
  size?: Size;
  fill?: string;
  type: IconType;
  [key: string]: any;
};

const Icon: SFC<IconProps> = ({
  type,
  cursor = 'default',
  size = 24,
  ...otherProps
}) => {
  let IconComponent = type;

  if (typeof type === 'string') {
    const BuiltInIcon = (icons as Icons)[type];

    if (!BuiltInIcon) {
      throw new Error('Built-in icon does not exists!');
    }

    IconComponent = BuiltInIcon;
  }

  return (
    <IconWrapper cursor={cursor} size={size} {...otherProps}>
      <IconComponent />
    </IconWrapper>
  );
};

export default Icon;
