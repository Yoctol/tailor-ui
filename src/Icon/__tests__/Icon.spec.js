import React from 'react';

import mountWithTheme from 'test/mountWithTheme';

import Icon from '..';

describe('Icon', () => {
  it('should render correctly', () => {
    const component = mountWithTheme(<Icon type="understood" />);

    expect(component).toMatchSnapshot();
  });
});
