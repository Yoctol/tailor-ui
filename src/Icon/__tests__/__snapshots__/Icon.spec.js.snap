// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon should render correctly 1`] = `
.c0 {
  display: inline-block;
  line-height: 1;
  cursor: default;
}

.c0 svg {
  vertical-align: middle;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
  fill: #373737;
  width: 24px;
  height: 24px;
}

<i
  class="c0"
  cursor="default"
  fill="gray.2"
  size="24"
>
  <svg
    viewBox="0 0 48 48"
    xmlns="https://www.w3.org/2000/svg"
    xmlns:xlink="https://www.w3.org/1999/xlink"
  >
    <title>
      Understood Icon
    </title>
    <defs>
      <path
        d="M44.19 27.947l-3.358-6.718c-.175-11.063-10.26-19.586-21.619-17.38-7.11 1.382-12.775 7.101-14.096 14.232-.93 5.02.187 10.013 3.144 14.059 2.415 3.303 3.744 7.266 3.745 11.159 0 .663.537 1.201 1.2 1.201h16.817c.664 0 1.201-.538 1.201-1.201v-3.14c0-.344.28-.624.625-.624h5.381a3.608 3.608 0 0 0 3.604-3.604v-3.603h.65c1.056 0 2.019-.537 2.574-1.436.556-.899.605-2 .133-2.945zm-2.175 1.682a.616.616 0 0 1-.532.297h-1.85c-.664 0-1.201.537-1.201 1.2v4.806c0 .662-.54 1.2-1.202 1.2h-5.38a3.03 3.03 0 0 0-3.028 3.028v1.938H14.371c-.245-4.004-1.704-8.002-4.17-11.376-2.564-3.507-3.53-7.84-2.722-12.203 1.143-6.169 6.043-11.116 12.192-12.31 9.86-1.916 18.76 5.51 18.76 15.308 0 .187.044.37.127.537l3.484 6.968a.616.616 0 0 1-.027.607zM22.816 10.707c-5.96 0-10.81 4.85-10.81 10.81 0 5.961 4.85 10.811 10.81 10.811 5.961 0 10.81-4.85 10.81-10.81 0-5.962-4.849-10.811-10.81-10.811zm0 19.219c-4.636 0-8.408-3.772-8.408-8.409 0-4.636 3.772-8.408 8.408-8.408 4.636 0 8.408 3.772 8.408 8.408 0 4.637-3.772 8.409-8.408 8.409zm5.654-11.66a1.2 1.2 0 0 0-1.698 0l-5.157 5.156-2.754-2.754a1.201 1.201 0 1 0-1.699 1.699l3.604 3.603a1.197 1.197 0 0 0 1.698 0l6.006-6.006c.47-.469.47-1.23 0-1.698z"
        id="understood_a"
      />
    </defs>
    <g
      fill-rule="evenodd"
    >
      <mask
        fill="#fff"
        id="understood_b"
      >
        <use
          xlink:href="#understood_a"
        />
      </mask>
      <use
        fill="#000"
        fill-rule="nonzero"
        xlink:href="#understood_a"
      />
      <g
        mask="url(#understood_b)"
        stroke="#0C0D19"
      >
        <path
          d="M-.5-.5h49v49h-49z"
        />
      </g>
    </g>
  </svg>
</i>
`;
